lib/utils/aws/credentials.go

```diff
@@ -0,0 +1,132 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package aws
+
+import (
+	"context"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/client"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/credentials/stscreds"
+	"github.com/gravitational/trace"
+	"github.com/jonboulle/clockwork"
+	"github.com/sirupsen/logrus"
+
+	"github.com/gravitational/teleport/lib/utils"
+)
+
+// GetCredentialsRequest is the request for obtains STS credentials.
+type GetCredentialsRequest struct {
+	// Provider is the user session used to create the STS client.
+	Provider client.ConfigProvider
+	// Expiry is session expiry to be requested.
+	Expiry time.Time
+	// SessionName is the session name to be requested.
+	SessionName string
+	// RoleARN is the role ARN to be requested.
+	RoleARN string
+	// ExternalID is the external ID to be requested, if not empty.
+	ExternalID string
+}
+
+// CredentialsGetter defines an interface for obtaining STS credentials.
+type CredentialsGetter interface {
+	// Get obtains STS credentials.
+	Get(ctx context.Context, request GetCredentialsRequest) (*credentials.Credentials, error)
+}
+
+type credentialsGettter struct {
+}
+
+// NewCredentialsGetter returns a new CredentialsGetter.
+func NewCredentialsGetter() CredentialsGetter {
+	return &credentialsGettter{}
+}
+
+// Get obtains STS credentials.
+func (g *credentialsGettter) Get(_ context.Context, request GetCredentialsRequest) (*credentials.Credentials, error) {
+	logrus.Debugf("Creating STS session %q for %q.", request.SessionName, request.RoleARN)
+	return stscreds.NewCredentials(request.Provider, request.RoleARN,
+		func(cred *stscreds.AssumeRoleProvider) {
+			cred.RoleSessionName = request.SessionName
+			cred.Expiry.SetExpiration(request.Expiry, 0)
+
+			if request.ExternalID != "" {
+				cred.ExternalID = aws.String(request.ExternalID)
+			}
+		},
+	), nil
+}
+
+// CachedCredentialsGetterConfig is the config for creating a CredentialsGetter that caches credentials.
+type CachedCredentialsGetterConfig struct {
+	// Getter is the CredentialsGetter for obtaining the STS credentials.
+	Getter CredentialsGetter
+	// CacheTTL is the cache TTL.
+	CacheTTL time.Duration
+	// Clock is used to control time.
+	Clock clockwork.Clock
+}
+
+// SetDefaults sets default values for CachedCredentialsGetterConfig.
+func (c *CachedCredentialsGetterConfig) SetDefaults() {
+	if c.Getter == nil {
+		c.Getter = NewCredentialsGetter()
+	}
+	if c.CacheTTL <= 0 {
+		// Minimum AssumeRole duration is 900 seconds. And leave some room for overhead.
+		c.CacheTTL = 800 * time.Second
```

Comment:
The ~13 minutes TTL feels like a lot. There shouldn't be any performance problems if we were to lower this to something like 60 seconds?

Reply:
sounds good. Will do one minute as other places. I don't think there will be a performance problem. We are currently doing this for every single API call. Even caching for a second will save a lot when using things like terraform.